### PR TITLE
[PKX-010] Search Pokemon functionality

### PR DESCRIPTION
**Pull request by giantiyan**

### CODE REVIEW

File: lib/features/home_page/home_page.dart [hunk 1]
```diff
@@ -3,18 +3,21 @@ import 'package:flutter/material.dart';
...
 import 'package:pokedex/utilities/colors.dart';
 
 class HomePage extends StatelessWidget {
   const HomePage({
-    required this.onCancel,
+    required this.onSearch,
     required this.onFilter,
+    required this.onCancel,
     this.pokemon,
     this.isVisible,
```
Comment (ryle-roasa) [inline lib/features/home_page/home_page.dart]: could you rename this to a more specific variable name

File: lib/features/home_page/home_page.dart [hunk 2]
```diff
@@ -83,23 +56,22 @@ class HomePage extends StatelessWidget {
...
-          if (isVisible == true)(
-              FloatingActionButton(
-                heroTag: 'cancel',
-                onPressed: onCancel,
-                backgroundColor: red,
-                child: const Icon(Icons.cancel),
-              )
-          ),
+          if (isVisible == true)
+            (FloatingActionButton(
```
Comment (ryle-roasa) [inline lib/features/home_page/home_page.dart]: no need for the parentheses

File: lib/features/home_page/home_page_vm.dart [hunk 3]
```diff
@@ -9,40 +9,55 @@ import 'package:pokedex/state/app_state.dart';
...
-        pokemon: state.pokemon,
         isVisible: state.isVisible,
         pageState: _getPageState(),
       );
 
   UnionPageState<List<PokemonModel>?> _getPageState() {
     if (state.wait.isWaitingFor(getPokemonListKey)) {
       return const UnionPageState.loading();
+    } else if (state.searchedPokemon?.isNotEmpty == true) {
+      return UnionPageState(state.searchedPokemon);
```
Comment (ryle-roasa) [inline lib/features/home_page/home_page_vm.dart:23]:
```
else if (state.searchedPokemons?.isEmpty == true) {
  return UnionPageState([]);
}
```
for when searchedPokemon is empty

File: lib/features/home_page/home_page_vm.dart [hunk 4]
```diff
@@ -9,40 +9,55 @@ import 'package:pokedex/state/app_state.dart';
...
     } else if (state.pokemon?.isNotEmpty == true) {
       return UnionPageState(state.pokemon);
     } else {
       return const UnionPageState.error('Home Page Error Message');
     }
   }
 
+  void _onSearchPokemons(query) {
```
Comment (ryle-roasa) [inline lib/features/home_page/home_page_vm.dart:33]: nice

File: lib/features/home_page/widgets/search_bar.dart [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+class _SearchBarState extends State<SearchBar> {
+  final _controller = TextEditingController();
+  bool? isIconVisible;
+
+  @override
+  Widget build(BuildContext context) {
+    return Container(
+      height: 46,
+      width: 220,
```
Comment (ryle-roasa) [inline lib/features/home_page/widgets/search_bar.dart]: could you move these two magic numbers to constants file and name it `searchBarheight` or something like that

File: lib/features/home_page/widgets/search_bar.dart [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+      child: TextField(
+        textAlignVertical: TextAlignVertical.top,
+        controller: _controller,
+        decoration: InputDecoration(
+          isDense: true,
+          fillColor: white,
+          filled: true,
+          prefixIcon: Icon(Icons.search),
+          border: OutlineInputBorder(
+            borderRadius: BorderRadius.circular(25.0),
```
Comment (ryle-roasa) [inline lib/features/home_page/widgets/search_bar.dart]: remove trailing comma

File: lib/features/home_page/widgets/search_bar.dart [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+          ),
+          suffixIcon: _controller.text.isNotBlank
+              ? IconButton(
+                  icon: Icon(Icons.clear),
+                  onPressed: () {
+                    _controller.clear();
+                    widget.onSearch('');
+                    FocusScope.of(context).requestFocus(FocusNode());
```
Comment (ryle-roasa) [inline lib/features/home_page/widgets/search_bar.dart:38]: could you explain what is this for? surprised

File: lib/state/actions/home_page_actions.dart [hunk 8]
```diff
@@ -23,6 +25,28 @@ class GetPokemonAction extends LoadingAction {
...
+class SearchPokemonAction extends ReduxAction<AppState> {
+  SearchPokemonAction(this.searchedPokemon);
+
+  final List<PokemonModel>? searchedPokemon;
+
+  @override
+  AppState reduce() => state.copyWith(searchedPokemon: searchedPokemon);
+}
+
+class ClearSearchPokemonAction extends ReduxAction<AppState> {
```
Comment (ryle-roasa) [inline lib/state/actions/home_page_actions.dart:48]: documentation

File: lib/features/home_page/widgets/search_bar.dart [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+  final ValueChanged<String> onSearch;
+
+  @override
+  State<SearchBar> createState() => _SearchBarState();
+}
+
+class _SearchBarState extends State<SearchBar> {
+  final _controller = TextEditingController();
+  bool? isIconVisible;
```
Comment (ryle-roasa) [inline lib/features/home_page/widgets/search_bar.dart]: i think you can remove this

File: lib/features/home_page/widgets/search_bar.dart (hunk 7 above)
Comment (giantiyan) [reply lib/features/home_page/widgets/search_bar.dart:38]: Hides the keyboard and removes the focus on the textfield when the 'x' icon is pressed.

**Review by ryle-roasa — APPROVED** [review]

**Review by SaturnoMark — APPROVED** [review]

**Pull request merged**